openedx/core/djangoapps/catalog/utils.py

```diff
@@ -93,16 +112,118 @@ def get_program_types():
+    )
+
+    program_instructors_dict = {}
+    program_instructors_list = cache.get(cache_key, [])
+    if program_instructors_list:
+        return program_instructors_list
+
+    module_store = modulestore()
+    for course_run_key in _get_program_course_run_keys(program):
+        course_descriptor = module_store.get_course(course_run_key)
+        if course_descriptor:
+            course_instructors = getattr(course_descriptor, 'instructor_info', {})
+            # Deduplicate program instructors using instructor name
+            program_instructors_dict.update(
+                {instructor.get('name'): instructor for instructor in course_instructors.get('instructors', [])}
+            )
+
+    program_instructors_list = program_instructors_dict.values()
+    cache.set(cache_key, program_instructors_list)
+
+    return program_instructors_list
+
+
+def _get_program_course_run_keys(program):
     """
-    Join program type logos with programs of corresponding type.
+    Returns a list of course keys for each course run in the given program.
+
+    Arguments:
+        program (dict): representing the program.
+
+    Returns:
+        list of CourseKey, for each course run key in the program.
     """
-    programs_list = get_programs()
-    program_types = get_program_types()
+    keys = []
+    for course in program.get('courses', []):  # pylint: disable=E1101
+        for course_run in course.get('course_runs', []):
+            keys.append(CourseKey.from_string(course_run.get('key')))
+    return keys
 
-    type_logo_map = {program_type['name']: program_type['logo_image'] for program_type in program_types}
 
-    for program in programs_list:
-        program['logo_image'] = type_logo_map[program['type']]
+def get_program_with_type_and_instructors(marketing_slug):
+    """
+    Return the program with full program type data and instructor info.
+    The instructor info added here is stored on the course module and can be
+    edited in Studio.
 
-    return programs_list
+    Arguments:
+        marketing_slug (string): The program marketing slug.
+
+    Returns:
+        dict, representing the program.
+    """
+    program_with_type_and_instructors = None
+    programs = get_programs()
+
+    # Find the program using the marketing slug.
+    program = next((program for program in programs if program.get('marketing_slug') == marketing_slug), None)
+    if program:
+        # A call to get_programs() with no UUID makes a call
+        # to the catalog service's programs list endpoint which
+        # uses a serializer that excludes some program data
+        # that we need here.
+        # Get the program by UUID in order to get the fully serialized program.
+        fully_serialized_program = next(iter(get_programs(uuid=program.get('uuid'))), None)
+
+        if fully_serialized_program:
+            # Deep copy the program dict here so we are not adding
+            # the type and instructors to the cached object.
```

Comment:
I don't think we should assume that memcached is the cache backend. This came up when running tests because we use an in-memory cache when using test settings.